### PR TITLE
Add a gradle flag to build using downloaded node/npm

### LINKED ISSUE #4667: Use the gradle node plugin to download node, npm, ...
Just like with the frontend-maven-plugin, we should setup gradle to install node, npm and such from a gradle task.
This way, the only requirement to build a JHipster project is JDK8.

### PR DESCRIPTION
**Pull request by cbornet**
* Add a nodeDl property to use a downloaded node/npm (as with frontend-maven-plugin). Default is deactivated.
* Add a workaround for gradle-node-plugin/issues/134 (only works for npm so even if yarn was used at project generation, it fallsback to npm if nodeDl flag is on)

Fix #4667

### CODE REVIEW

**Comment by cbornet** [comment]
With this you can build with a bare openjdk image
```
./gradlew clean
./gradlew npmInstall -PnodeDl
./gradlew test -PnodeDl
./gradlew gulp_test -PnodeDl
./gradlew bootRepackage -x test -Pprod -PnodeDl
```

File: generators/server/templates/_build.gradle [hunk 1]
```diff
@@ -15,10 +16,7 @@ buildscript {
         classpath "org.springframework.boot:spring-boot-gradle-plugin:${spring_boot_version}"
         classpath "org.springframework.build.gradle:propdeps-plugin:0.0.7"
         <%_ if(!skipClient) { _%>
-        classpath "com.moowork.gradle:gradle-node-plugin:0.14"
-            <%_ if(clientFw === 'angular1') { _%>
-        classpath "com.moowork.gradle:gradle-gulp-plugin:0.13"
```
Comment (deepu105) [inline generators/server/templates/_build.gradle]: is gulp included in the node plugin?

File: generators/server/templates/gradle/_yeoman.gradle [hunk 2]
```diff
@@ -1,12 +1,20 @@
...
 
-task bower(type: Exec) {
-    if (Os.isFamily(Os.FAMILY_WINDOWS)) {
-        commandLine 'cmd', '/c', 'bower', 'install'
-    }else{
-        commandLine 'bower', 'install'
-    }
+task bower(type: NodeTask) {
+    description = "Installs dependencies using Bower"
+    script = file("${project.projectDir}/node_modules/bower/bin/bower")
```
Comment (deepu105) [inline generators/server/templates/gradle/_yeoman.gradle]: I dont think the path is same in windows/unix can you please check?

File: generators/server/templates/_build.gradle [hunk 3]
```diff
@@ -369,7 +373,29 @@ task wrapper(type: Wrapper) {
 
 task stage(dependsOn: 'bootRepackage') {
 }
+<% if(!skipClient) { %>
+if (project.hasProperty('nodeDl')) {
```
Comment (deepu105) [inline generators/server/templates/_build.gradle]: im not happy with the property name can we have something better may be like `nodeInstall` or `localNode` or something

File: generators/server/templates/_build.gradle (hunk 1 above)
Comment (cbornet) [inline generators/server/templates/_build.gradle]: yes

File: generators/server/templates/_build.gradle (hunk 3 above)
Comment (cbornet) [inline generators/server/templates/_build.gradle]: OK for nodeInstall

File: generators/server/templates/gradle/_yeoman.gradle (hunk 2 above)
Comment (deepu105) [inline generators/server/templates/gradle/_yeoman.gradle]: ya but the I saw somebody added a os check on our new webpack task on one of the gradle file, were based on os the bin path was different

Comment (cbornet) [inline generators/server/templates/gradle/_yeoman.gradle]: Yes. On Linux node_modules/.bin/webpack is a symlink to node_modules/webpack/bin/webpack.js. So IMO this platform check was useless and both could use node_modules/webpack/bin/webpack.js directly.

**Comment by deepu105** [comment]
OH ok then we can remove it. DO you wanna do it as part of this PR?


Thanks & Regards,
Deepu

On Wed, Dec 14, 2016 at 7:24 PM, Christophe Bornet <notifications@github.com

**Comment by cbornet** [comment]
Yep.

**Comment by cbornet** [comment]
@deepu105 Is your review OK ? Once this is merge, I will update #4600

**Pull request merged**